### ISSUE TITLE
Unable to cook Potions of Healing with Alchemy Pot

### ISSUE THREAD

**Issue by ansiyuan** [00-Evan/shattered-pixel-dungeon#57]
According to `Heap.transmute()` (lines 424 and 425), with more successfully cooked Potions of Healing, the Alchemy Pot is more likely to randomize its output:
```java
while (potion instanceof PotionOfHealing && Random.Int(15) - Dungeon.limitedDrops.cookingHP.count >= 0)
	potion = Generator.random( Generator.Category.POTION );
```
But this loop prevents Potions of Healing from ever being output, since `Dungeon.limitedDrops.cookingHP.count` is initialized to 0 and `Random.Int(15) - Dungeon.limitedDrops.cookingHP.count >= 0` always evaluates to `true`.
This issue can be fixed by reversing the comparison in the loop condition:
```java
Random.Int(15) - Dungeon.limitedDrops.cookingHP.count < 0
```
Which is equivalent to:
```java
Random.Int(15) < Dungeon.limitedDrops.cookingHP.count
```

**Comment by 00-Evan** [comment]
The order of operations will not complete in this way, as java will attempt to resolve arithmetic before any comparison operators.

So Random.Int(15) - Dungeon.limitedDrops.cookingHP.count will calculate into a number which is then compared to 0.

If it were to resolve in the way you suggest java would then attempt to calculate the result of Random.Int(15) - false, which obviously cannot happen in a strongly typed language like Java.

Still, I certainly do agree that the code could be a bit more easily readable, will change it to 

`Random.Int(15) >= Dungeon.limitedDrops.cookingHP.count`

Which is logical equivalent but easier to understand.

**Comment by ansiyuan** [comment]
Oops! There was a mistake in my original description of the issue. It's correct now.

Here's why `Random.Int(15) >= Dungeon.limitedDrops.cookingHP.count` doesn't fix the problem. `Dungeon.limitedDrops.cookingHP.count` starts at 0, so the expression above becomes `Random.Int(15) >= 0`, which evaluates to `true`. The loop will execute until the output potion type is something other than Healing, which means that `Dungeon.limitedDrops.cookingHP.count` will stay at 0.

**Comment by 00-Evan** [comment]
Ah, I see, the problem is the inversion, it should be true when we want to ditch the health potion and false otherwise, not the inverse.

So we would want `Random.Int(15) < Dungeon.limitedDrops.cookingHP.count`

**Comment by ansiyuan** [comment]
Exactly. Keep up the good work!

**Comment by 00-Evan** [comment]
This issue has been addressed in the latest beta.

You can opt in to receive betas from google play here: https://play.google.com/apps/testing/com.shatteredpixel.shatteredpixeldungeon

Once the source is released this ticket will be closed with the commit referenced.

**Comment by 00-Evan** [comment]
fixed in commit: 0845c41a447404ea8bf5e56480694fda8c2b28ea
